src/vm/fieldmarshaler.cpp

```diff
@@ -1806,16 +1829,22 @@ VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
             if (!(alignmentRequirement == 1 ||
                      alignmentRequirement == 2 ||
                      alignmentRequirement == 4 ||
-                  alignmentRequirement == 8))
+                  alignmentRequirement == 8 ||
+                  alignmentRequirement == 16 ||
+                  alignmentRequirement == 32))
             {
                 COMPlusThrowHR(COR_E_INVALIDPROGRAM, BFA_METADATA_CORRUPT);
             }
-            
-            alignmentRequirement = min(alignmentRequirement, packingSize);
-            
+
+            // Only take the smaller of the two if we don't contain a SIMD hardware intrinsic type
```

Comment:
This  and other similar special cases look suspect to me.

Reply:
Could you elaborate?

Reply:
The existing logic will fail to align the types properly (for the default packing scenario) unless we also change `DEFAULT_PACKING_SIZE` to the 32.

However, these types also have a special ABI rule that they must always be aligned, so this check is required to ensure that a user specified smaller packing is ignored for these types:
> Like the Intel386 architecture, the AMD64 architecture in general does not require all data accesses to be properly aligned. Misaligned data accesses are slower than aligned accesses but otherwise behave identically. The only exceptions are that __m128, __m256 and __m512 must always be aligned properly.

Reply:
ARM64 has a similar requirement:
> If the argument is an HFA, an HVA, a Quad-precision Floating-point or Short Vector Type then the NSAA is rounded up to the larger of 8 or the Natural Alignment of the argument’s type.

Reply:
Also, as a note: MSVC will not let you violate the `__declspec(align(x))` requirements of these types (even if you explicitly mark the containing types as `pack(1)` and `align(1)`).

However, Clang and GCC will.

I went with the current behavior because it most closely follows the ABI requirements.

Reply:
Would you rather this be handled by explicitly tracking whether the type, or any subtypes, contains one of the SIMD HWIntrinsic types?

Currently, this is reliant on those being specially handled in the VM (which was similar to how we handle alignment for the existing primitive types, when an ABI required different packing/alignment from the default).

Reply:
@davidwrighton and @MichalStrehovsky , could you please help Tanner  to figure out the right way to modify the field layout algorithm for Vector types?

Reply:
> Would you rather this be handled by explicitly tracking whether the type, or any subtypes, contains one of the SIMD HWIntrinsic types?

That would feel the most natural to me - a property of the type (same as e.g. `Align8` of `IsHFA`), and not something we guess based on magic numbers.

Reply:
@MichalStrehovsky, @jkotas: After talking with @sdmaclea below on the ARM32/ARM64 ABI requirements and looking at what the GCC/Clang compilers do despite the ABI requirements (including on x86/x64), I believe the appropriate change is to drop this special handling (which was guaranteeing ABI compliance in spite of attempted user override).

Instead, the default packing should be changed to the largest supported packing of a "fundamental data type" (this change will make it `32`, instead of the current `8`).

The existing `min(alignment, packing)` logic will ensure that any data structures that don't contain these larger fundamental data types will still have the same packing as they do today (8, 4, 2, or 1).

User-specified smaller packing will also work on these types due to the existing logic.

Does that sound reasonable?

Reply:
> the default packing should be changed to the largest supported packing

Is this going to change layout for any existing structures that do not contain VectorXX fields?

Reply:
No, the logic (both prior to and after the proposed change) will be to pick `min(alignment, packing)`, where alignment is the natural alignment of a fundamental type or the largest alignment requirement of all members of a structure.

So the overall packing for a structure will never go above the largest alignment requirement of its members.

This is why:
```
[StructLayout(LayoutKind.Sequential, Pack = 16, Size = 16)]
struct MyStruct { }
```

Has a packing of 1.

Reply:
So why do we have `DEFAULT_PACKING_SIZE` constant at all? Can this constant be `0x80000000` ?

Reply:
> So why do we have DEFAULT_PACKING_SIZE constant at all.

Not sure. It is only used in 1 location today (although there are a few other places in fieldmarshaler where it should probably be used). We also use a different constant for doing things with AutoLayout.

> Can this constant be 0x80000000

There is currently a soft limit of 128 (in the methodtablebuilder).

Reply:
The existing rules are documented here: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.structlayoutattribute.pack(v=vs.110).aspx

The handling of packing (where it is ignored when larger than the alignment of its largest element) is actually why this PR is needed for the HWIntrinsic types. If it wasn't ignored, users could specify the correct packing for their type and have it "just work". However, changing that at this stage would be a breaking change.